liblumen_alloc/src/erts/term/map.rs

```diff
@@ -55,6 +55,14 @@ impl Map {
         self.value.len()
     }
 
+    pub fn update(&self, key: Term, value: Term) -> Option<HashMap<Term, Term>> {
+        let mut map = self.value.clone();
```

Comment:
@KronicDeth I haven't really had the time to look into the runtime yet, but have you guys considered using a [HAMT-based](https://en.wikipedia.org/wiki/Hash_array_mapped_trie) HashMap? It's an immutable implementation where clones are cheap as most of the memory is shared -- it's what BEAM uses internally (and Clojure and a bunch of other language implementations).

I've used [im::HashMap](https://docs.rs/im/13.0.0/im/hashmap/struct.HashMap.html) before, but I'm not sure how much work it would be to make it`no_std` compatible.

Reply:
We tried using `im`, it caused bugs.  Using HashBrown is safe if inefficient.  We assumed that `im`’s optimizations for when the ref count is 1 was incompatible with how terms are boxed and it was taking too much time to fix, so this worked.  You are welcome to figure out the borrow issues if you want to try again.  We were likely to just implement our own persistent hash later that works better with our allocators.

Reply:
Even without `im::HashMap` this could be more efficient by checking for the key first and not cloning when it won't succeed, but PR functionally works and alter PRs can make it more efficient.